packages/api/amplify_api/test/ws/web_socket_connection_test.dart

```diff
@@ -137,42 +149,109 @@ void main() {
+      );
+
+      expect(
+        hubEvents,
+        emitsInOrder(
+          [
+            connectingHubEvent,
+            connectedHubEvent,
+            connectingHubEvent,
+            connectedHubEvent,
+          ],
+        ),
+      );
+
+      await assertWebSocketConnected(connection, subscriptionRequest.id);
+
+      connection.channel!.sink.add(jsonEncode(mockSubscriptionEvent));
+
+      fakePlatform.controller.sink.add(ConnectivityResult.none);
+      fakePlatform.controller.sink.add(ConnectivityResult.wifi);
+      fakePlatform.controller.sink.add(ConnectivityResult.none);
+      fakePlatform.controller.sink.add(ConnectivityResult.wifi);
+      fakePlatform.controller.sink.add(ConnectivityResult.none);
+      fakePlatform.controller.sink.add(ConnectivityResult.wifi);
+      fakePlatform.controller.sink.add(ConnectivityResult.none);
+      fakePlatform.controller.sink.add(ConnectivityResult.wifi);
+      fakePlatform.controller.sink.add(ConnectivityResult.none);
+      fakePlatform.controller.sink.add(ConnectivityResult.wifi);
+
+      await expectLater(connection.reconnectPending, completes);
+
+      await assertWebSocketConnected(connection, subscriptionRequest.id);
+    });
+
     test(
-      'should reconnect when network turns on/off',
-      () async {
-        final dataCompleter = Completer<String>();
-        final subscription = getWebSocketConnection().subscribe(
-          subscriptionRequest,
-          () {},
-        );
-
-        await assertWebSocketConnected(connection, subscriptionRequest.id);
-
-        final streamSub = subscription.listen(
-          expectAsync1(
-            (event) => dataCompleter.complete(event.data),
-          ),
-        );
-
-        connection.channel!.sink.add(jsonEncode(mockSubscriptionEvent));
-
-        fakePlatform.controller.sink.add(ConnectivityResult.none);
-
-        expect(
-          hubEvents,
-          emitsInOrder(
-            [
-              connectingHubEvent,
-              connectedHubEvent,
-              connectingHubEvent,
-            ],
-          ),
-        );
-
-        addTearDown(streamSub.cancel);
-      },
-      skip: 'todo(equartey): Implement reconnection logic tests',
-    );
+        'should reconnect after 15 seconds when appsync ping fails multiple times',
+        () async {
+      getWebSocketConnection().subscribe(
+        subscriptionRequest,
+        expectAsync0(() {}),
+      );
+
+      expect(
+        hubEvents,
+        emitsInOrder(
+          [
+            connectingHubEvent,
+            connectedHubEvent,
+            connectingHubEvent,
+            connectedHubEvent,
+          ],
+        ),
+      );
+
+      await assertWebSocketConnected(connection, subscriptionRequest.id);
+      fakePlatform.controller.sink.add(ConnectivityResult.wifi);
+
+      pleaseFail = true;
+
+      await expectLater(pingFailed.future, completes);
+
+      await Future<void>.delayed(const Duration(seconds: 15));
```

Comment:
Not necessary here but might be fun to explore fake_async in the future to avoid literal delays https://pub.dev/packages/fake_async